File: src/test/java/no/nav/bidrag/beregn/bidragsevne/BidragsevneberegningTest.java
package no.nav.bidrag.beregn.bidragsevne;

import static java.util.Collections.emptyList;
import static java.util.Collections.singletonList;
import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.jupiter.api.Assertions.assertEquals;

import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.List;
import no.nav.bidrag.beregn.TestUtil;
import no.nav.bidrag.beregn.bidragsevne.beregning.BidragsevneberegningImpl;
import no.nav.bidrag.beregn.bidragsevne.bo.GrunnlagBeregningPeriodisert;
import no.nav.bidrag.beregn.bidragsevne.bo.Inntekt;
import no.nav.bidrag.beregn.felles.SjablonUtil;
import no.nav.bidrag.beregn.felles.bo.Sjablon;
import no.nav.bidrag.beregn.felles.bo.SjablonInnhold;
import no.nav.bidrag.beregn.felles.enums.BostatusKode;
import no.nav.bidrag.beregn.felles.enums.InntektType;
import no.nav.bidrag.beregn.felles.enums.SaerfradragKode;
import no.nav.bidrag.beregn.felles.enums.SjablonInnholdNavn;
import no.nav.bidrag.beregn.felles.enums.SjablonTallNavn;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

@DisplayName("BidragsevneBeregningTest")
class BidragsevneberegningTest {

  private final List<Sjablon> sjablonListe = TestUtil.byggSjabloner();

  @Test
  void beregn() {

    ArrayList<Inntekt> inntekter = new ArrayList<>();

    BidragsevneberegningImpl bidragsevneberegning = new BidragsevneberegningImpl();

    // Tester beregning med ulike inntekter
    inntekter.add(new Inntekt(InntektType.LONN_SKE, BigDecimal.valueOf(1000000)));
    GrunnlagBeregningPeriodisert grunnlagBeregningPeriodisert
        = new GrunnlagBeregningPeriodisert(inntekter, 1, BostatusKode.ALENE, BigDecimal.ONE,
        SaerfradragKode.INGEN, sjablonListe);
    assertEquals(BigDecimal.valueOf(31859),
        bidragsevneberegning.beregn(grunnlagBeregningPeriodisert).getResultatEvneBelop());

    inntekter.set(0, new Inntekt(InntektType.LONN_SKE, BigDecimal.valueOf(520000)));
    GrunnlagBeregningPeriodisert grunnlagBeregningPeriodisert2
        = new GrunnlagBeregningPeriodisert(inntekter, 1, BostatusKode.ALENE, BigDecimal.ONE,
        SaerfradragKode.INGEN, sjablonListe);
    assertEquals(BigDecimal.valueOf(8322),
        bidragsevneberegning.beregn(grunnlagBeregningPeriodisert2).getResultatEvneBelop());
    assertEquals(BigDecimal.valueOf(10833),
        bidragsevneberegning.beregn(grunnlagBeregningPeriodisert2).getResultat25ProsentInntekt());

    inntekter.set(0, new Inntekt(InntektType.LONN_SKE, BigDecimal.valueOf(666000)));
    GrunnlagBeregningPeriodisert grunnlagBeregningPeriodisert3
        = new GrunnlagBeregningPeriodisert(inntekter, 1, BostatusKode.ALENE, BigDecimal.valueOf(3),
        SaerfradragKode.INGEN, sjablonListe);
    assertEquals(BigDecimal.valueOf(8424),
        bidragsevneberegning.beregn(grunnlagBeregningPeriodisert3).getResultatEvneBelop());

    // Test på at beregnet bidragsevne blir satt til 0 når evne er negativ
    inntekter.set(0, new Inntekt(InntektType.LONN_SKE, BigDecimal.valueOf(100000)));
    GrunnlagBeregningPeriodisert grunnlagBeregningPeriodisert4
        = new GrunnlagBeregningPeriodisert(inntekter, 1, BostatusKode.MED_ANDRE, BigDecimal.ONE,
        SaerfradragKode.HELT, sjablonListe);
    assertEquals(BigDecimal.ZERO,
        bidragsevneberegning.beregn(grunnlagBeregningPeriodisert4).getResultatEvneBelop());

    // Test at fordel skatteklasse 2 ikke legges til beregnet evne når skatteklasse = 1
    inntekter.set(0, new Inntekt(InntektType.LONN_SKE, BigDecimal.valueOf(666000)));
    sjablonListe.set(0, new Sjablon(SjablonTallNavn.FORDEL_SKATTEKLASSE2_BELOP.getNavn(), emptyList(),
        singletonList(new SjablonInnhold(SjablonInnholdNavn.SJABLON_VERDI.getNavn(), BigDecimal.valueOf(12000)))));

    GrunnlagBeregningPeriodisert grunnlagBeregningPeriodisert5
        = new GrunnlagBeregningPeriodisert(inntekter, 1, BostatusKode.ALENE, BigDecimal.valueOf(3),
        SaerfradragKode.INGEN, sjablonListe);
    assertEquals(BigDecimal.valueOf(8424),
        bidragsevneberegning.beregn(grunnlagBeregningPeriodisert5).getResultatEvneBelop());

    // Test at fordel skatteklasse 2 legges til beregnet evne når skatteklasse = 2
    inntekter.set(0, new Inntekt(InntektType.LONN_SKE, BigDecimal.valueOf(666000)));
    sjablonListe.set(0, new Sjablon(SjablonTallNavn.FORDEL_SKATTEKLASSE2_BELOP.getNavn(), emptyList(),
        singletonList(new SjablonInnhold(SjablonInnholdNavn.SJABLON_VERDI.getNavn(), BigDecimal.valueOf(12000)))));

    GrunnlagBeregningPeriodisert grunnlagBeregningPeriodisert6
        = new GrunnlagBeregningPeriodisert(inntekter, 2, BostatusKode.ALENE, BigDecimal.valueOf(3),
        SaerfradragKode.INGEN, sjablonListe);
    assertEquals(BigDecimal.valueOf(9424),
        bidragsevneberegning.beregn(grunnlagBeregningPeriodisert6).getResultatEvneBelop());

    // Test at personfradrag skatteklasse 2 brukes hvis skatteklasse 2 er angitt
    sjablonListe.set(0, new Sjablon(SjablonTallNavn.FORDEL_SKATTEKLASSE2_BELOP.getNavn(), emptyList(),
        singletonList(new SjablonInnhold(SjablonInnholdNavn.SJABLON_VERDI.getNavn(), BigDecimal.ZERO))));

    sjablonListe.set(1, new Sjablon(SjablonTallNavn.PERSONFRADRAG_KLASSE2_BELOP.getNavn(), emptyList(),
        singletonList(new SjablonInnhold(SjablonInnholdNavn.SJABLON_VERDI.getNavn(), BigDecimal.valueOf(24000)))));

    GrunnlagBeregningPeriodisert grunnlagBeregningPeriodisert7
        = new GrunnlagBeregningPeriodisert(inntekter, 2, BostatusKode.ALENE, BigDecimal.valueOf(3),
        SaerfradragKode.INGEN, sjablonListe);
    assertEquals(BigDecimal.valueOf(7923),
        bidragsevneberegning.beregn(grunnlagBeregningPeriodisert7).getResultatEvneBelop());


    // Test av halvt særfradrag
    GrunnlagBeregningPeriodisert grunnlagBeregningPeriodisert8
        = new GrunnlagBeregningPeriodisert(inntekter, 1, BostatusKode.ALENE, BigDecimal.valueOf(3),
        SaerfradragKode.HALVT, sjablonListe);
    assertEquals(BigDecimal.valueOf(8965),
        bidragsevneberegning.beregn(grunnlagBeregningPeriodisert8).getResultatEvneBelop());

    // Test av bostatus MED_FLERE
    GrunnlagBeregningPeriodisert grunnlagBeregningPeriodisert9
        = new GrunnlagBeregningPeriodisert(inntekter, 1, BostatusKode.MED_ANDRE, BigDecimal.valueOf(3),
        SaerfradragKode.HALVT, sjablonListe);
    assertEquals(BigDecimal.valueOf(14253),
        bidragsevneberegning.beregn(grunnlagBeregningPeriodisert9).getResultatEvneBelop());

  }

  @Test
  void beregnMinstefradrag() {

    ArrayList<Inntekt> inntekter = new ArrayList<>();
    inntekter.add(new Inntekt(InntektType.LONN_SKE, BigDecimal.valueOf(200000)));

    BidragsevneberegningImpl bidragsevneberegning = new BidragsevneberegningImpl();

    GrunnlagBeregningPeriodisert grunnlagBeregningPeriodisert
        = new GrunnlagBeregningPeriodisert(inntekter, 1, BostatusKode.ALENE, BigDecimal.ONE,
        SaerfradragKode.HELT, sjablonListe);
    System.out.println(bidragsevneberegning.beregnMinstefradrag(grunnlagBeregningPeriodisert,
        SjablonUtil.hentSjablonverdi(sjablonListe, SjablonTallNavn.MINSTEFRADRAG_INNTEKT_BELOP),
        SjablonUtil.hentSjablonverdi(sjablonListe, SjablonTallNavn.MINSTEFRADRAG_INNTEKT_PROSENT)));
    assertThat(bidragsevneberegning.beregnMinstefradrag(grunnlagBeregningPeriodisert,
        SjablonUtil.hentSjablonverdi(sjablonListe, SjablonTallNavn.MINSTEFRADRAG_INNTEKT_BELOP),
        SjablonUtil.hentSjablonverdi(sjablonListe, SjablonTallNavn.MINSTEFRADRAG_INNTEKT_PROSENT)))
        .isEqualTo(BigDecimal.valueOf(62000));

    inntekter.set(0, new Inntekt(InntektType.LONN_SKE, BigDecimal.valueOf(1000000)));
    GrunnlagBeregningPeriodisert grunnlagBeregningPeriodisert2
        = new GrunnlagBeregningPeriodisert(inntekter, 1, BostatusKode.ALENE, BigDecimal.ONE,
        SaerfradragKode.HELT, sjablonListe);
    System.out.println(bidragsevneberegning.beregnMinstefradrag(grunnlagBeregningPeriodisert2,
        SjablonUtil.hentSjablonverdi(sjablonListe, SjablonTallNavn.MINSTEFRADRAG_INNTEKT_BELOP),
        SjablonUtil.hentSjablonverdi(sjablonListe, SjablonTallNavn.MINSTEFRADRAG_INNTEKT_PROSENT)));
    assertThat(bidragsevneberegning.beregnMinstefradrag(grunnlagBeregningPeriodisert2,
        SjablonUtil.hentSjablonverdi(sjablonListe, SjablonTallNavn.MINSTEFRADRAG_INNTEKT_BELOP),
        SjablonUtil.hentSjablonverdi(sjablonListe, SjablonTallNavn.MINSTEFRADRAG_INNTEKT_PROSENT)))
        .isEqualTo(BigDecimal.valueOf(87450));
  }

  @Test
  void beregnSkattetrinnBelop() {

    BidragsevneberegningImpl bidragsevneberegning = new BidragsevneberegningImpl();

    ArrayList<Inntekt> inntekter = new ArrayList<>();
    inntekter.add(new Inntekt(InntektType.LONN_SKE, BigDecimal.valueOf(666000)));

    GrunnlagBeregningPeriodisert grunnlagBeregningPeriodisert
        = new GrunnlagBeregningPeriodisert(inntekter, 1, BostatusKode.ALENE, BigDecimal.ONE,
        SaerfradragKode.HELT, sjablonListe);
    assertEquals(BigDecimal.valueOf(1400+16181+3465+0),
        bidragsevneberegning.beregnSkattetrinnBelop(grunnlagBeregningPeriodisert));

    inntekter.set(0, new Inntekt(InntektType.LONN_SKE, BigDecimal.valueOf(174600)));
    GrunnlagBeregningPeriodisert grunnlagBeregningPeriodisert2
        = new GrunnlagBeregningPeriodisert(inntekter, 1, BostatusKode.ALENE, BigDecimal.ONE,
        SaerfradragKode.HELT, sjablonListe);
    assertEquals(BigDecimal.ZERO,
        bidragsevneberegning.beregnSkattetrinnBelop(grunnlagBeregningPeriodisert2));

    inntekter.set(0, new Inntekt(InntektType.LONN_SKE, BigDecimal.valueOf(250000)));
    GrunnlagBeregningPeriodisert grunnlagBeregningPeriodisert3
        = new GrunnlagBeregningPeriodisert(inntekter, 1, BostatusKode.ALENE, BigDecimal.ONE,
        SaerfradragKode.HELT, sjablonListe);
    assertEquals(BigDecimal.valueOf(1315),
        bidragsevneberegning.beregnSkattetrinnBelop(grunnlagBeregningPeriodisert3));
  }

  @Test
  void TestFraJohn() {

    ArrayList<Inntekt> inntekter = new ArrayList<>();

    BidragsevneberegningImpl bidragsevneberegning = new BidragsevneberegningImpl();

    // Tester beregning med ulike inntekter
    inntekter.add(new Inntekt(InntektType.LONN_SKE, BigDecimal.valueOf(300000)));
    GrunnlagBeregningPeriodisert grunnlagBeregningPeriodisert
        = new GrunnlagBeregningPeriodisert(inntekter, 1, BostatusKode.ALENE, BigDecimal.ZERO,
        SaerfradragKode.HELT, sjablonListe);
    assertEquals(BigDecimal.valueOf(1217),
        bidragsevneberegning.beregn(grunnlagBeregningPeriodisert).getResultatEvneBelop());

  }
}

File: src/test/kotlin/no/nav/bidrag/beregn/bidragsevne/BidragsevneBeregningTest.kt
package no.nav.bidrag.beregn.bidragsevne

import no.nav.bidrag.beregn.TestUtil.BARN_I_HUSSTAND_REFERANSE
import no.nav.bidrag.beregn.TestUtil.BOSTATUS_REFERANSE
import no.nav.bidrag.beregn.TestUtil.INNTEKT_REFERANSE
import no.nav.bidrag.beregn.TestUtil.SAERFRADRAG_REFERANSE
import no.nav.bidrag.beregn.TestUtil.SKATTEKLASSE_REFERANSE
import no.nav.bidrag.beregn.TestUtil.byggSjablonPeriodeListe
import no.nav.bidrag.beregn.TestUtil.byggSjabloner
import no.nav.bidrag.beregn.bidragsevne.beregning.BidragsevneBeregning.Companion.getInstance
import no.nav.bidrag.beregn.bidragsevne.bo.BarnIHusstand
import no.nav.bidrag.beregn.bidragsevne.bo.Bostatus
import no.nav.bidrag.beregn.bidragsevne.bo.GrunnlagBeregning
import no.nav.bidrag.beregn.bidragsevne.bo.Inntekt
import no.nav.bidrag.beregn.bidragsevne.bo.Saerfradrag
import no.nav.bidrag.beregn.bidragsevne.bo.Skatteklasse
import no.nav.bidrag.beregn.felles.bo.Periode
import no.nav.bidrag.beregn.felles.bo.Sjablon
import no.nav.bidrag.beregn.felles.bo.SjablonInnhold
import no.nav.bidrag.beregn.felles.bo.SjablonPeriode
import no.nav.bidrag.beregn.felles.util.SjablonUtil.hentSjablonverdi
import no.nav.bidrag.beregn.felles.util.SjablonUtil.hentTrinnvisSkattesats
import no.nav.bidrag.domene.enums.beregning.Særfradragskode
import no.nav.bidrag.domene.enums.person.Bostatuskode
import no.nav.bidrag.domene.enums.sjablon.SjablonInnholdNavn
import no.nav.bidrag.domene.enums.sjablon.SjablonNavn
import no.nav.bidrag.domene.enums.sjablon.SjablonTallNavn
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Assertions.assertAll
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.function.Executable
import java.math.BigDecimal
import java.time.LocalDate

@DisplayName("Test av beregning av bidragsevne")
internal class BidragsevneBeregningTest {
    private val sjablonListe = byggSjabloner()
    private var sjablonPeriodeListe = byggSjablonPeriodeListe().toMutableList()
    private val bidragsevneBeregning = getInstance()

    @Test
    @DisplayName("Skal beregne bidragsevne med inntekt 1000000")
    fun skalBeregneBidragsevneMedInntekt1000000() {
        val grunnlagBeregning =
            GrunnlagBeregning(
                inntektListe = listOf(Inntekt(referanse = INNTEKT_REFERANSE, type = "LONN_SKE", belop = BigDecimal.valueOf(1000000))),
                skatteklasse = Skatteklasse(referanse = SKATTEKLASSE_REFERANSE, skatteklasse = 1),
                bostatus = Bostatus(referanse = BOSTATUS_REFERANSE, kode = Bostatuskode.ALENE),
                barnIHusstand = BarnIHusstand(referanse = BARN_I_HUSSTAND_REFERANSE, antallBarn = 1.0),
                saerfradrag = Saerfradrag(referanse = SAERFRADRAG_REFERANSE, kode = Særfradragskode.INGEN),
                sjablonListe = sjablonPeriodeListe,
            )

        val (belop) = bidragsevneBeregning.beregn(grunnlagBeregning)

        assertThat(BigDecimal.valueOf(31859)).isEqualTo(belop)
    }

    @Test
    @DisplayName("Skal beregne bidragsevne med inntekt 520000")
    fun skalBeregneBidragsevneMedInntekt520000() {
        val grunnlagBeregning =
            GrunnlagBeregning(
                inntektListe = listOf(Inntekt(referanse = INNTEKT_REFERANSE, type = "LONN_SKE", belop = BigDecimal.valueOf(520000))),
                skatteklasse = Skatteklasse(referanse = SKATTEKLASSE_REFERANSE, skatteklasse = 1),
                bostatus = Bostatus(referanse = BOSTATUS_REFERANSE, kode = Bostatuskode.ALENE),
                barnIHusstand = BarnIHusstand(referanse = BARN_I_HUSSTAND_REFERANSE, antallBarn = 1.0),
                saerfradrag = Saerfradrag(referanse = SAERFRADRAG_REFERANSE, kode = Særfradragskode.INGEN),
                sjablonListe = sjablonPeriodeListe,
            )

        val (belop, inntekt25Prosent) = bidragsevneBeregning.beregn(grunnlagBeregning)

        assertAll(
            Executable { assertThat(BigDecimal.valueOf(8322)).isEqualTo(belop) },
            Executable { assertThat(BigDecimal.valueOf(10833)).isEqualTo(inntekt25Prosent) },
        )
    }

    @Test
    @DisplayName("Skal beregne bidragsevne med inntekt 600000")
    fun skalBeregneBidragsevneMedInntekt600000() {
        val grunnlagBeregning =
            GrunnlagBeregning(
                inntektListe = listOf(Inntekt(referanse = INNTEKT_REFERANSE, type = "LONN_SKE", belop = BigDecimal.valueOf(600000))),
                skatteklasse = Skatteklasse(referanse = SKATTEKLASSE_REFERANSE, skatteklasse = 1),
                bostatus = Bostatus(referanse = BOSTATUS_REFERANSE, kode = Bostatuskode.ALENE),
                barnIHusstand = BarnIHusstand(referanse = BARN_I_HUSSTAND_REFERANSE, antallBarn = 0.0),
                saerfradrag = Saerfradrag(referanse = SAERFRADRAG_REFERANSE, kode = Særfradragskode.HELT),
                sjablonListe = sjablonPeriodeListe,
            )

        val (belop) = bidragsevneBeregning.beregn(grunnlagBeregning)

        assertThat(BigDecimal.valueOf(17617)).isEqualTo(belop)
    }

    @Test
    @DisplayName("Skal beregne bidragsevne med inntekt 666000")
    fun skalBeregneBidragsevneMedInntekt666000() {
        val grunnlagBeregning =
            GrunnlagBeregning(
                inntektListe = listOf(Inntekt(referanse = INNTEKT_REFERANSE, type = "LONN_SKE", belop = BigDecimal.valueOf(666000))),
                skatteklasse = Skatteklasse(referanse = SKATTEKLASSE_REFERANSE, skatteklasse = 1),
                bostatus = Bostatus(referanse = BOSTATUS_REFERANSE, kode = Bostatuskode.ALENE),
                barnIHusstand = BarnIHusstand(referanse = BARN_I_HUSSTAND_REFERANSE, antallBarn = 3.0),
                saerfradrag = Saerfradrag(referanse = SAERFRADRAG_REFERANSE, kode = Særfradragskode.INGEN),
                sjablonListe = sjablonPeriodeListe,
            )

        val (belop) = bidragsevneBeregning.beregn(grunnlagBeregning)

        assertThat(BigDecimal.valueOf(8424)).isEqualTo(belop)
    }

    @Test
    @DisplayName("Skal beregne bidragsevne lik 0 når evne er negativ")
    fun skalBeregneBidragsevneLik0NaarEvneErNegativ() {
        val grunnlagBeregning =
            GrunnlagBeregning(
                inntektListe = listOf(Inntekt(referanse = INNTEKT_REFERANSE, type = "LONN_SKE", belop = BigDecimal.valueOf(100000))),
                skatteklasse = Skatteklasse(referanse = SKATTEKLASSE_REFERANSE, skatteklasse = 1),
                bostatus = Bostatus(referanse = BOSTATUS_REFERANSE, kode = Bostatuskode.IKKE_MED_FORELDER),
                barnIHusstand = BarnIHusstand(referanse = BARN_I_HUSSTAND_REFERANSE, antallBarn = 1.0),
                saerfradrag = Saerfradrag(referanse = SAERFRADRAG_REFERANSE, kode = Særfradragskode.HELT),
                sjablonListe = sjablonPeriodeListe,
            )

        val (belop) = bidragsevneBeregning.beregn(grunnlagBeregning)

        assertThat(BigDecimal.ZERO).isEqualTo(belop)
    }

    @Test
    @DisplayName("Skal ikke legge fordel skatteklasse 2 til beregnet evne for skatteklasse lik 1")
    fun skalIkkeLeggeFordelSkatteklasse2TilBeregnetEvneForSkatteklasseLik1() {
        sjablonPeriodeListe.add(
            SjablonPeriode(
                sjablonPeriode = Periode(datoFom = LocalDate.parse("2017-01-01"), datoTil = LocalDate.parse("9999-12-31")),
                sjablon =
                Sjablon(
                    navn = SjablonTallNavn.FORDEL_SKATTEKLASSE2_BELØP.navn,
                    nokkelListe = emptyList(),
                    innholdListe =
                    listOf(
                        SjablonInnhold(navn = SjablonInnholdNavn.SJABLON_VERDI.navn, verdi = BigDecimal.valueOf(12000)),
                    ),
                ),
            ),
        )
        val grunnlagBeregning =
            GrunnlagBeregning(
                inntektListe = listOf(Inntekt(referanse = INNTEKT_REFERANSE, type = "LONN_SKE", belop = BigDecimal.valueOf(666000))),
                skatteklasse = Skatteklasse(referanse = SKATTEKLASSE_REFERANSE, skatteklasse = 1),
                bostatus = Bostatus(referanse = BOSTATUS_REFERANSE, kode = Bostatuskode.ALENE),
                barnIHusstand = BarnIHusstand(referanse = BARN_I_HUSSTAND_REFERANSE, antallBarn = 3.0),
                saerfradrag = Saerfradrag(referanse = SAERFRADRAG_REFERANSE, kode = Særfradragskode.INGEN),
                sjablonListe = sjablonPeriodeListe,
            )

        val (belop) = bidragsevneBeregning.beregn(grunnlagBeregning)

        assertThat(BigDecimal.valueOf(8424)).isEqualTo(belop)
    }

    @Test
    @DisplayName("Skal legge fordel skatteklasse 2 til beregnet evne for skatteklasse lik 2")
    fun skalLeggeFordelSkatteklasse2TilBeregnetEvneForSkatteklasseLik2() {
        sjablonPeriodeListe.add(
            0,
            SjablonPeriode(
                sjablonPeriode = Periode(datoFom = LocalDate.parse("2017-01-01"), datoTil = LocalDate.parse("9999-12-31")),
                sjablon =
                Sjablon(
                    navn = SjablonTallNavn.FORDEL_SKATTEKLASSE2_BELØP.navn,
                    nokkelListe = emptyList(),
                    innholdListe =
                    listOf(
                        element = SjablonInnhold(navn = SjablonInnholdNavn.SJABLON_VERDI.navn, verdi = BigDecimal.valueOf(12000)),
                    ),
                ),
            ),
        )
        val grunnlagBeregning =
            GrunnlagBeregning(
                inntektListe = listOf(Inntekt(referanse = INNTEKT_REFERANSE, type = "LONN_SKE", belop = BigDecimal.valueOf(666000))),
                skatteklasse = Skatteklasse(referanse = SKATTEKLASSE_REFERANSE, skatteklasse = 2),
                bostatus = Bostatus(referanse = BOSTATUS_REFERANSE, kode = Bostatuskode.ALENE),
                barnIHusstand = BarnIHusstand(referanse = BARN_I_HUSSTAND_REFERANSE, antallBarn = 3.0),
                saerfradrag = Saerfradrag(referanse = SAERFRADRAG_REFERANSE, kode = Særfradragskode.INGEN),
                sjablonListe = sjablonPeriodeListe,
            )

        val (belop) = bidragsevneBeregning.beregn(grunnlagBeregning)

        assertThat(BigDecimal.valueOf(9424)).isEqualTo(belop)
    }

    @Test
    @DisplayName("Skal bruke halvt barn pga delt bosted")
    fun skalBrukeHalvtBarnPgaDeltBosted() {
        val grunnlagBeregning =
            GrunnlagBeregning(
                inntektListe = listOf(Inntekt(referanse = INNTEKT_REFERANSE, type = "LONN_SKE", belop = BigDecimal.valueOf(666000))),
                skatteklasse = Skatteklasse(referanse = SKATTEKLASSE_REFERANSE, skatteklasse = 2),
                bostatus = Bostatus(referanse = BOSTATUS_REFERANSE, kode = Bostatuskode.IKKE_MED_FORELDER),
                barnIHusstand = BarnIHusstand(referanse = BARN_I_HUSSTAND_REFERANSE, antallBarn = 0.5),
                saerfradrag = Saerfradrag(referanse = SAERFRADRAG_REFERANSE, kode = Særfradragskode.INGEN),
                sjablonListe = sjablonPeriodeListe,
            )

        val (belop) = bidragsevneBeregning.beregn(grunnlagBeregning)

        assertThat(BigDecimal.valueOf(23314)).isEqualTo(belop)
    }

    @Test
    @DisplayName("Skal bruke personfradrag skatteklasse 2 hvis skatteklasse 2 er angitt")
    fun skalBrukePersonfradragSkatteklasse2HvisSkatteklasse2ErAngitt() {
        sjablonPeriodeListe.add(
            0,
            SjablonPeriode(
                sjablonPeriode = Periode(datoFom = LocalDate.parse("2017-01-01"), datoTil = LocalDate.parse("9999-12-31")),
                sjablon =
                Sjablon(
                    navn = SjablonTallNavn.PERSONFRADRAG_KLASSE2_BELØP.navn,
                    nokkelListe = emptyList(),
                    innholdListe =
                    listOf(
                        SjablonInnhold(navn = SjablonInnholdNavn.SJABLON_VERDI.navn, verdi = BigDecimal.valueOf(24000)),
                    ),
                ),
            ),
        )
        val grunnlagBeregning =
            GrunnlagBeregning(
                inntektListe = listOf(Inntekt(referanse = INNTEKT_REFERANSE, type = "LONN_SKE", belop = BigDecimal.valueOf(666000))),
                skatteklasse = Skatteklasse(referanse = SKATTEKLASSE_REFERANSE, skatteklasse = 2),
                bostatus = Bostatus(referanse = BOSTATUS_REFERANSE, kode = Bostatuskode.ALENE),
                barnIHusstand = BarnIHusstand(referanse = BARN_I_HUSSTAND_REFERANSE, antallBarn = 3.0),
                saerfradrag = Saerfradrag(referanse = SAERFRADRAG_REFERANSE, kode = Særfradragskode.INGEN),
                sjablonListe = sjablonPeriodeListe,
            )

        val (belop) = bidragsevneBeregning.beregn(grunnlagBeregning)

        assertThat(BigDecimal.valueOf(7923)).isEqualTo(belop)
    }

    @Test
    @DisplayName("Skal bruke halvt særfradrag")
    fun skalBrukeHalvtSaerfradrag() {
        val grunnlagBeregning =
            GrunnlagBeregning(
                inntektListe = listOf(Inntekt(referanse = INNTEKT_REFERANSE, type = "LONN_SKE", belop = BigDecimal.valueOf(666000))),
                skatteklasse = Skatteklasse(referanse = SKATTEKLASSE_REFERANSE, skatteklasse = 1),
                bostatus = Bostatus(referanse = BOSTATUS_REFERANSE, kode = Bostatuskode.ALENE),
                barnIHusstand = BarnIHusstand(referanse = BARN_I_HUSSTAND_REFERANSE, antallBarn = 3.0),
                saerfradrag = Saerfradrag(referanse = SAERFRADRAG_REFERANSE, kode = Særfradragskode.HALVT),
                sjablonListe = sjablonPeriodeListe,
            )

        val (belop) = bidragsevneBeregning.beregn(grunnlagBeregning)

        assertThat(BigDecimal.valueOf(8965)).isEqualTo(belop)
    }

    @Test
    @DisplayName("Skal bruke bostatus med andre")
    fun skalBrukeBostatusMedAndre() {
        val grunnlagBeregning =
            GrunnlagBeregning(
                inntektListe = listOf(Inntekt(referanse = INNTEKT_REFERANSE, type = "LONN_SKE", belop = BigDecimal.valueOf(666000))),
                skatteklasse = Skatteklasse(referanse = SKATTEKLASSE_REFERANSE, skatteklasse = 1),
                bostatus = Bostatus(referanse = BOSTATUS_REFERANSE, kode = Bostatuskode.IKKE_MED_FORELDER),
                barnIHusstand = BarnIHusstand(referanse = BARN_I_HUSSTAND_REFERANSE, antallBarn = 3.0),
                saerfradrag = Saerfradrag(referanse = SAERFRADRAG_REFERANSE, kode = Særfradragskode.HALVT),
                sjablonListe = sjablonPeriodeListe,
            )

        val (belop) = bidragsevneBeregning.beregn(grunnlagBeregning)

        assertThat(BigDecimal.valueOf(14253)).isEqualTo(belop)
    }

    @Test
    @DisplayName("Skal hente sjablonverdier for trinnvis skattesats")
    fun skalhenteSjablonverdierForTrinnvisSkattesats() {
        val sjablonVerdiSkattesats =
            hentSjablonverdi(
                sjablonListe = sjablonListe,
                sjablonTallNavn = SjablonTallNavn.SKATTESATS_ALMINNELIG_INNTEKT_PROSENT,
            )
        val sortertTrinnvisSkattesatsListe =
            hentTrinnvisSkattesats(sjablonListe = sjablonListe, sjablonNavn = SjablonNavn.TRINNVIS_SKATTESATS)

        assertAll(
            Executable { assertThat(sjablonVerdiSkattesats).isEqualTo(BigDecimal.valueOf(22)) },
            Executable { assertThat(sortertTrinnvisSkattesatsListe.size).isEqualTo(4) },
            Executable { assertThat(sortertTrinnvisSkattesatsListe[0].inntektGrense).isEqualTo(BigDecimal.valueOf(180800)) },
            Executable { assertThat(sortertTrinnvisSkattesatsListe[0].sats).isEqualTo(BigDecimal.valueOf(1.9)) },
        )
    }
}
